Update "iframe with interactive elements is not excluded from tab-order" rule to manage inert iframe elements

giacomo-petri opened the pull request:
This pull request manages inert iframe elements in [iframe with interactive elements is not excluded from tab-order](https://www.w3.org/WAI/standards-guidelines/act/rules/akn7bn/proposed/) rule, excluding them from the applicability of the rule.

Updates:
- edited applicability excluding inert iframe elements;
- added definition of _inert_ in applicability;
- added definition of _modal_ (modal dialog);
- added 2 inapplicable examples:
  - iframe with inert attribute
  - modal which overlaps the iframe, making the iframe inoperable



Closes issue(s): https://github.com/act-rules/act-rules.github.io/issues/1965

Need for Call for Review:
This will require a 2 weeks Call for Review

---

## Pull Request Etiquette

### **When creating PR:**

- [ ] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [ ] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [ ] Add yourself (and co-authors) as "Assignees" for PR.
- [ ] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

giacomo-petri committed:
Update programmatically-determined-link-context.md

giacomo-petri committed:
Merge branch 'act-rules:develop' into develop

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update programmatically-determined-link-context.md

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Create modal-dialog

giacomo-petri committed:
Rename modal-dialog to modal-dialog.md

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

carlosapaduarte committed:
Merge branch 'develop' into giacomo-petri-iframe-with-interactive-element-patch

carlosapaduarte commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 
 An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
```
I think we can have this as a definition also, to make it reusable in other (future?) rules

carlosapaduarte replied:
And probably it would be better to have a definite of inert element, not just inert iframe

Jym77 replied:
I agree, we should have a separate definition of "inert element".

carlosapaduarte commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
+- it has an `inert` [attribute value][] of true; or
+- an element which is not the iframe itself and that is not [contained](#akn7bn:contain) in it behaves as a [modal][], making the iframe inoperable.
```
What does "behaves as a modal" mean?
And do we want to tie this to the behaviour or the state? If there is a modal open, the iframe is inert and not applicable, but if the modal is closed, then the iframe is applicable. In both states, the component behaves as a modal, right?

Jym77 replied:
See https://html.spec.whatwg.org/multipage/interaction.html#modal-dialogs-and-inert-subtrees

Jym77 replied:
I think we can probably refer directly to the [HTML definition of inert](https://html.spec.whatwg.org/multipage/interaction.html#inert).
It doesn't refer explicitly to modals, but [being blocked by a modal](https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog) causes most of the document to be inert, so we can mention that.

If we want to have our own definition of inert, we should probably refer to "being blocked by a modal", since it is objective and unambiguous, and a common definition from higher spec.
This may leave room for custom modals where somebody uses a `div` and a ton of Javascript to turn it into a modal manually. We might be OK enough by having an assumption about people not being that crazy… (i.e. I do not know how frequent it is for devs to re-code modal behaviour on their own rather than relying on native semantics…)

carlosapaduarte commented on pages/glossary/modal-dialog.md:
```diff
@@ -0,0 +1,16 @@
+  - DOM tree
+---
+
+A modal dialog is a type of user interface element that appears on top of the main content of a [html web page][] or application and requires the user to take some action or make a decision before they can continue interacting with the program.
```
I don't think "type of user interface" is objective.
Same for "main content". 
What is the "program" you mention?

I think this needs to focus on the element preventing interaction with all other elements that are not their children. 

This definition really needs to be objective, because it is used in the applicability of the rule.

dan-tripp-siteimprove replied:
```suggestion
A modal dialog is an element which, until it is closed, prevents user interaction with any content on the web page that is not a descendant element of the modal.  Visually, a modal often (but not necessarily) appears on top of all other content of the [html web page][] and visually darkens or otherwise obscures that other content.
```
Trying to put Carlos' advice into action.

dan-tripp-siteimprove commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
+- it has an `inert` [attribute value][] of true; or
+- an element which is not the iframe itself and that is not [contained](#akn7bn:contain) in it behaves as a [modal][], making the iframe inoperable.
```
```suggestion
- an element which is not the iframe itself and that is not [contained](#akn7bn:contain) in it makes the iframe and all of its contents [inoperable][].  (For example: a modal.)
```
Does that improve the situation at all?

WilcoFiers replied:
What I meant by modal dialogs was the actual `<dialog>` element with `showModal()` called on it. That programmatically makes all other content on the page inert. I think you could define it like this:

> The element is in a page with a `dialog` element with the [`is-modal`](https://html.spec.whatwg.org/multipage/interactive-elements.html#is-modal) flag set to `true`, and is not a descendant of that `dialog` element.


I'm not sure we should try for a generic modal definition. I don't know how you'd do that in a way that's objective.

Jym77 replied:
Then, we are indeed going for [blocked by a modal](https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog):
> A [Document](https://html.spec.whatwg.org/multipage/dom.html#document) document is blocked by a modal dialog subject if subject is the topmost [dialog](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element) element in document's [top layer](https://fullscreen.spec.whatwg.org/#top-layer). While document is so blocked, every node that is [connected](https://dom.spec.whatwg.org/#connected) to document, with the exception of the subject element and its [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) descendants, must become [inert](https://html.spec.whatwg.org/multipage/interaction.html#inert).

> subject can additionally become [inert](https://html.spec.whatwg.org/multipage/interaction.html#inert) via the [inert](https://html.spec.whatwg.org/multipage/interaction.html#the-inert-attribute) attribute, but only if specified on subject itself (i.e., subject escapes inertness of ancestors); subject's [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) descendants can become [inert](https://html.spec.whatwg.org/multipage/interaction.html#inert) in a similar fashion.

> Note: The [dialog](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element) element's [showModal()](https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-showmodal) method causes this mechanism to trigger, by [adding](https://fullscreen.spec.whatwg.org/#top-layer-add) the [dialog](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element) element to its [node document](https://dom.spec.whatwg.org/#concept-node-document)'s [top layer](https://fullscreen.spec.whatwg.org/#top-layer).

dan-tripp-siteimprove commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -122,10 +126,32 @@ This `iframe` element contains a link that is not part of its [sequential focus
 [attribute value]: #attribute-value 'Definition of Attribute Value'
 [container document]: https://html.spec.whatwg.org/#bc-container-document 'HTML browsing context container document, 2020/12/18'
 [document]: https://html.spec.whatwg.org/multipage/dom.html#document 'HTML definition of document'
 [flattened tabindex-ordered focus navigation scope]: https://html.spec.whatwg.org/multipage/interaction.html#flattened-tabindex-ordered-focus-navigation-scope 'HTML - Living Standard, 2022/07/08'
```
```suggestion
[flattened tabindex-ordered focus navigation scope]: https://html.spec.whatwg.org/multipage/interaction.html#flattened-tabindex-ordered-focus-navigation-scope 'HTML - Living Standard, 2022/07/08'
[inoperable]: https://www.w3.org/TR/wai-aria/#dfn-operable
```

WilcoFiers replied:
This looks outdated?

WilcoFiers commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
+- it has an `inert` [attribute value][] of true; or
```
The attribute value doesn't matter. It just has to be present. Plus this should apply to ancestors too. Inert disables the entire subtree.

Jym77 replied:
> The attribute value doesn't matter. It just has to be present. 

As per [our definition of attribute value](https://github.com/act-rules/act-rules.github.io/blob/develop/pages/glossary/attribute-value.md), boolean attribute (like `inert`) have an "attribute value" of `true` if the attribute is present.

WilcoFiers commented on pages/glossary/modal-dialog.md:
```diff
@@ -0,0 +1,16 @@
+---
+title: Modal dialog
```
I'm concerned this could be confused by the `<dialog>`'s `showModal()` method.

carlosapaduarte committed:
Merge branch 'develop' into giacomo-petri-iframe-with-interactive-element-patch

dan-tripp-siteimprove commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 An element is <dfn id="akn7bn:contain">contained</dfn> in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
+- it has an `inert` [attribute value][] of true; or
```
```suggestion
- it has an `inert` [attribute value][] of true; or
- one of its ancestor elements has an `inert` [attribute value][] of true; or
```

dan-tripp-siteimprove commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -28,13 +28,17 @@ acknowledgments:
 
+An `iframe` element is <dfn id="akn7bn:inert">inert</dfn> if:
+- it has an `inert` [attribute value][] of true; or
+- an element which is not the iframe itself and that is not [contained](#akn7bn:contain) in it behaves as a [modal][], making the iframe inoperable.
```
```suggestion
- it is on the same page as a `dialog` element which has its [`is-modal`](https://html.spec.whatwg.org/multipage/interactive-elements.html#is-modal) flag set to `true`, and is not a descendant of that `dialog` element.
```

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>

Jym77 committed:
Merge branch 'develop' into giacomo-petri-iframe-with-interactive-element-patch

Jym77 reviewed (dismissed):
To sum up, I think we can just use HTML definition of inert (instead of trying to rebuild it) + add some background description about `showModal` triggering "blocked by a modal" triggering inert + make the examples descriptions clear (this `iframe` is inert because it is blocked by the modal that has been brought to the top layer with `showModal`).

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update and rename modal-dialog.md to inert.md

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri commented:
Updates:
- added "inert" definition to be re-used;
- updated the definition of inert by using the "blocked by a modal" html definition;
- removed the modal definition completely (no longer needed);
- added example (inapplicable 6) of showModal() method.

I've kept the inapplicable example n. 7 with the custom modal. Let me know if it's better to remove it completely.

giacomo-petri committed:
Update spelling-ignore.yml

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update inert.md

bruce-usab reviewed (changes requested):
The name of the referenced rule `Iframe with interactive elements is not excluded from tab-order` is fine -- as is the content of the rule.

But the name of this PR seems to be a double negative?  It is very confusing!  And maybe it does not matter?

Please change the PR name.  If it were just "update iframe not focusable" I would have twice about the rule!

Jym77 commented:
> Title edited. The initial intent was to provide information about the specific update, avoiding a generic label such as "Update iframe rule", which doesn't help in filtering PR. Please, let me know if the new title makes sense.

As a rule of thumb, I feel it is good to include the full name of the updated rule in the PR title. Including the rule ID can also help tracking them across change of name (the id is the stable name). I usually like to frontload these, but we have no strict PR naming convention.
> "rule name" [rule ID]: verb something something.
e.g.
> "iframe with interactive elements is not excluded from tab-order" [akn7bn]: Manage inert iframe element

(I feel that "update rule" is a bit redundant in a PR title, since PR are mostly made to update rules)
(the current title is fine, no need to change it again)

dan-tripp-siteimprove commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -122,10 +122,61 @@ This `iframe` element contains a link that is not part of its [sequential focus
+
+#### Inapplicable Example 6
+
+Once "Privacy policy details" button is activated, the `iframe` element become [inert][] because of showModal() method that causes the `iframe` to be [blocked by a modal].
```
```suggestion
Once the "Privacy policy details" button is activated, the `iframe` element becomes [inert][] because of the showModal() method, which causes the `iframe` to be [blocked by a modal].
```

Jym77 replied:
I think we should make the example so that the modal is shown at load time.
Currently, examples do not require manipulation before testing. There are some steps in that direction as part of the "states" discussion (#1953, #2045).

In this case, we can simply call the `showModal` in an `onload` call to ensure this is testable out of the box. I think the example will still look "real" enough that people will understand why the modal is opened by default. (otherwise a more realstic case could be to have a cookies banner or region selection modal that is opened on page load and blocking interaction with the rest of the page).

Jym77 commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
```diff
@@ -122,10 +122,61 @@ This `iframe` element contains a link that is not part of its [sequential focus
+
+#### Inapplicable Example 7
+
+This `iframe` element is [inert][] because it is [blocked by a modal].
```
```suggestion
This `iframe` element is [inert][] because it is [blocked by a modal][].
```
This will also need an entry in the references list.

Jym77 replied:
It's technically not true. The HTML definition of "blocked by a modal" requires a `dialog` element to do the blocking, while this is done by a `div`.

Also, this actually doesn't works. Due to non-interference with the host language, `aria-modal` does not make the rest of the page inert. [ARIA for `aria-modal`](https://www.w3.org/TR/wai-aria-1.2/#aria-modal) states:
> When a modal element is displayed, authors SHOULD mark all other contents as inert (such as "inert subtrees" in HTML)

When testing (without the negative `tabindex`), I have no problem tabbing onto the link inside the `iframe`. I haven't tested whether screen reader consider it as inert, but Chrome doesn't.

Jym77 commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
````diff
@@ -122,10 +122,61 @@ This `iframe` element contains a link that is not part of its [sequential focus
+</div>
+<iframe tabindex="-1" srcdoc="<a href='/'>Home</a>"></iframe>
+```
+
 [attribute value]: #attribute-value 'Definition of Attribute Value'
````
```suggestion

[attribute value]: #attribute-value 'Definition of Attribute Value'
[blocked by a modal]: https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog
```

Jym77 commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
````diff
@@ -122,10 +122,61 @@ This `iframe` element contains a link that is not part of its [sequential focus
+This `iframe` element is [inert][] because of its own `inert` [attribute value][].
+
+```html
+<iframe inert srcdoc="<a href='/'>Home</a>"></iframe>
````
I think the `iframe` shuld have a negative `tabindex` under the rule of thumb "the Inapplicable example should look bad if it wasn't inapplicable". The reason is that in implemetation reports "Passed" and "Inapplicable" results are essentially equivalent, so to detect wrong Applicability in a tool, we need the tool to (wrongly) report "Failed" rather than to wrongly report "Passed".
```suggestion
<iframe tabindex="-1" inert srcdoc="<a href='/'>Home</a>"></iframe>
```

Jym77 commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
````diff
@@ -122,10 +122,61 @@ This `iframe` element contains a link that is not part of its [sequential focus
+Once "Privacy policy details" button is activated, the `iframe` element become [inert][] because of showModal() method that causes the `iframe` to be [blocked by a modal].
+
+```html
+<iframe srcdoc="<a href='/'>Home</a>"></iframe>
````
```suggestion
<iframe tabindex="-1" srcdoc="<a href='/'>Home</a>"></iframe>
```

Jym77 commented on pages/glossary/inert.md:
```diff
@@ -0,0 +1,15 @@
+  - DOM tree
+---
+
+An [HTML or SVG element][] is inert if:
+- it has an `inert` [attribute value][] of true; or
+- one of its ancestor elements has an `inert` [attribute value][] of true; or
+- it is [blocked by a modal](https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog).
```
All the references need to be defined in a references list at the bottom of the file.

giacomo-petri reviewed (commented):
suggestions added

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>
Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>

giacomo-petri committed:
Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

giacomo-petri committed:
Update inert.md

giacomo-petri commented:
- Accepted all the @Jym77's suggestions about adding the tabindex="-1" attribute to the iframe element;
- Updated the Failed example n. 6 with dialog opened `onload` instead of after user activation;
- Removed failed example n.7 with "custom modal" as not aligned with the rule;
- Updated references in "inert" definition.

Jym77 committed:
Merge branch 'develop' into giacomo-petri-iframe-with-interactive-element-patch

carlosapaduarte commented:
Call for review ends June 21

bruce-usab reviewed (approved):
Thanks for this!

From the list email, I had expected I would have concern about defining `inert`.  Instead, I was delighted to learn that it is a reasonably well defined attribute.

I don't think my own knowledge is too below average, so I recommend adding links for background reading.  Here are two I found useful:

- [Introducing inert](https://developer.chrome.com/articles/inert/)
- [MDN reference inert](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert)

Jym77 commented:
> I don't think my own knowledge is too below average, so I recommend adding links for background reading. Here are two I found useful:
> 
> * [Introducing inert](https://developer.chrome.com/articles/inert/)
> * [MDN reference inert](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert)

@bruce-usab Since ACT rules are intended to be published as official W3C documentation, we tend to avoid direct links (or citation) of commercial third parties, especially when they are themselves major players (we do not want to look like we favour one above the others). As much as we can, we'd rather stick with W3C documentation (HTML living standard, …)

Jym77 committed:
Merge branch 'develop' into giacomo-petri-iframe-with-interactive-element-patch

WilcoFiers commented on _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:
````diff
@@ -122,10 +122,42 @@ This `iframe` element contains a link that is not part of its [sequential focus
+Once the "Privacy policy details" button is activated, the `iframe` element becomes [inert][] because of the showModal() method, which causes the `iframe` to be [blocked by a modal].
+
+```html
+<iframe tabindex="-1" srcdoc="<a href='/'>Home</a>"></iframe>
+<div>
+    <button id="ppDetails">Privacy policy details</button>
+</div>
+<dialog id="ppDialog" aria-labelledby="dialogLabel">
+    <h2 id="dialogLabel">Privacy Policy</h2>
+    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in c ...
+    <button id="cancel">Cancel</button>
+</dialog>
+<script>    
+			const dialog = document.getElementById("ppDialog");
+			window.addEventListener('load', function(event) {
+    			dialog.showModal();
+			});
+</script>
````
I'm not entirely happy with this example. You can easily get this in an inaccessible state. Before onload the page is inaccessible, and after the modal is closed too. I also think that using "DOMContentLoaded" would be better, since most tools start testing when onload is triggered. This ensures the modal is already open when automated testing starts. I also added a title attribute to the iframe, and removed the duplicate ID. Lastly, I adjusted whitespace so it's what our formatter expects.

```suggestion
<iframe id="myFrame" title="Links" srcdoc="<a href='/'>Home</a>"></iframe>
<div>
	<button id="ppButton" onclick="openDialog()">Privacy policy details</button>
</div>
<dialog id="ppDialog" aria-labelledby="dialogLabel">
	<h2 id="dialogLabel">Privacy Policy</h2>
	<p>We store no data.</p>
	<button id="cancel" onclick="ppDialog.close()">Cancel</button>
</dialog>
<script>
	const openDialog = () => {
		ppDialog.showModal();
		myFrame.tabIndex = '-1'
	}
	ppDialog.addEventListener('close', () => myFrame.tabIndex = 0)
	window.addEventListener('DOMContentLoaded', openDialog);
</script>
```

WilcoFiers commented on pages/glossary/inert.md:
```diff
@@ -0,0 +1,19 @@
+
+An [HTML or SVG element][] is inert if:
+- it has an `inert` [attribute value][] of true; or
+- one of its ancestor elements has an `inert` [attribute value][] of true; or
```
We should indicate what tree to use when talking about ancestors.

```suggestion
- one of its ancestor elements in the [flat tree][] has an `inert` [attribute value][] of true; or
```

WilcoFiers commented on pages/glossary/inert.md:
```diff
@@ -0,0 +1,19 @@
+
+[HTML or SVG element]: #namespaced-element 'Definition of HTML or SVG element'
+[attribute value]: #attribute-value 'Definition of Attribute value'
+[blocked by a modal]: https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog
```
```suggestion
[blocked by a modal]: https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog
[flat tree]: https://drafts.csswg.org/css-scoping/#flat-tree 'Definition of flat tree'
```

WilcoFiers reviewed (changes requested):
Couple minor points. Don't expect any discussion necessary on these.

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

giacomo-petri commented:
This PR is in call for review for 2 weeks. Review ends Sept 29th.

WilcoFiers reviewed (approved):
Good stuff!
